### PR TITLE
[ARM,ELF] Fix access to dso_preemptable __stack_chk_guard with static relocation model

### LINKED ISSUE #64999: [LTO][PIC][ARM][SSP] extra indirection accessing __stack_chk_guard
After commit [e018cbf7208b](https://github.com/llvm/llvm-project/commit/e018cbf7208b3d34f18997ddee84c66cee32fb1b) we're seeing segfaults when accessing __stack_chk_guard when using LTO. I'm attaching a small test case (3 source files and small build script) that shows the problem). When compiling with -flto -fPIC and linking without any additional flags, we're seeing the following code in the function foo:
```   
201dc: e59f006c      ldr     r0, [pc, #0x6c]         @ 0x20250 <foo+0x80>
201e0: e5900000      ldr     r0, [r0]
201e4: e5900000      ldr     r0, [r0]
201e8: e50b000c      str     r0, [r11, #-0xc]
...
20250: 54 02 03 00   .word   0x00030254
```
And from the linker map file:
```
30254    30254        4     1                 __stack_chk_guard
```
So we're seeing an extra indirection, causing the segfaults.

If we're adding -pie to the link line, things work fine. We see:
```
   10280: e59f0074      ldr     r0, [pc, #0x74]         @ 0x102fc <foo+0x88>
   10284: e08f0000      add     r0, pc, r0
   10288: e5900000      ldr     r0, [r0]
   1028c: e5900000      ldr     r0, [r0]
    ...
   102fc: e4 00 01 00   .word   0x000100e4
```
with the GOT present in the executable.

Note that the the test case is not meant to execute anywhere. A rudimentary module (rt.c) supplies the missing definitions.
[stack_check.zip](https://github.com/llvm/llvm-project/files/12444315/stack_check.zip)

### PR DESCRIPTION
**Pull request by MaskRay**
The ELF code from https://reviews.llvm.org/D112811 emits LDRLIT_ga_pcrel
when `TM.isPositionIndependent()` but uses a different condition
`Subtarget.isGVIndirectSymbol(GV)` (aka dso_preemptable on ELF targets).
This would cause incorrect access for dso_preemptable
`__stack_chk_guard` with the static relocation model.

Regarding whether `__stack_chk_guard` gets the dso_local specifier,
https://reviews.llvm.org/D150841 switched to
`M.getDirectAccessExternalData()` (implied by "PIC Level") instead of
`TM.getRelocationModel() == Reloc::Static`.

The result is that when non-zero "PIC Level" is used with static
relocation model (e.g. -fPIE/-fPIC LTO compiles with -no-pie linking),
`__stack_chk_guard` accesses are incorrect.
```
        ldr     r0, .LCPI0_0
        ldr     r0, [r0]
        ldr     r0, [r0]   // incorrectly dereferences __stack_chk_guard
...
.LCPI0_0:
        .long   __stack_chk_guard
```

To fix this, for dso_preemptable `__stack_chk_guard`, emit a GOT PIC
code sequence like for -fpic using `LDRLIT_ga_pcrel`:
```
        ldr     r0, .LCPI0_0
.LPC0_0:
        add     r0, pc, r0
        ldr     r0, [r0]
        ldr     r0, [r0]
...
LCPI0_0:
.Ltmp0:
        .long   __stack_chk_guard(GOT_PREL)-((.LPC0_0+8)-.Ltmp0)
```

Technically, `LDRLIT_ga_abs` with `R_ARM_GOT_ABS` could be used, but
`R_ARM_GOT_ABS` does not have GNU or integrated assembler support.
(Note, `.LCPI0_0: .long __stack_chk_guard@GOT` produces an
`R_ARM_GOT_BREL`, which is not desired).

This patch fixes #6499 while not changing behavior for the following
configurations:
```
run arm.linux.nopic --target=arm-linux-gnueabi -fno-pic
run arm.linux.pie --target=arm-linux-gnueabi -fpie
run arm.linux.pic --target=arm-linux-gnueabi -fpic
run armv6.darwin.nopic --target=armv6-apple-darwin -fno-pic
run armv6.darwin.dynamicnopic --target=armv6-apple-darwin -mdynamic-no-pic
run armv6.darwin.pic --target=armv6-apple-darwin -fpic
run armv7.darwin.nopic --target=armv7-apple-darwin -mcpu=cortex-a8 -fno-pic
run armv7.darwin.dynamicnopic --target=armv7-apple-darwin -mcpu=cortex-a8 -mdynamic-no-pic
run armv7.darwin.pic --target=armv7-apple-darwin -mcpu=cortex-a8 -fpic
run arm64.darwin.pic --target=arm64-apple-darwin
```

### CODE REVIEW

File: llvm/lib/Target/ARM/ARMInstrInfo.cpp [hunk 1]
```diff
@@ -104,11 +104,12 @@ void ARMInstrInfo::expandLoadStackGuard(MachineBasicBlock::iterator MI) const {
   const GlobalValue *GV =
       cast<GlobalValue>((*MI->memoperands_begin())->getValue());
 
-  if (!Subtarget.useMovt() || Subtarget.isGVInGOT(GV)) {
```
Comment (MaskRay) [inline llvm/lib/Target/ARM/ARMInstrInfo.cpp:107]: `isGVInGOT` is defined as

```
bool ARMSubtarget::isGVInGOT(const GlobalValue *GV) const {
  return isTargetELF() && TM.isPositionIndependent() &&
         !TM.shouldAssumeDSOLocal(*GV->getParent(), GV);
}
```

I believe the function name is misleading and `TM.isPositionIndependent() &&` probably should be removed, so I avoid using it.
However, removing `TM.isPositionIndependent() &&` from the function uncovers other problems, so I do not touch `ARMSubtarget::isGVInGOT`.

Comment (MaskRay) [reply llvm/lib/Target/ARM/ARMInstrInfo.cpp:107]: I agree that we can technically use `LDRLIT_ga_abs` to load the GOT entry address but 
ARMExpandPseudoInsts.cpp:2640 does not use GOT for `LDRLIT_ga_abs`.

Comment (MaskRay) [reply llvm/lib/Target/ARM/ARMInstrInfo.cpp:107]: Technically with dso_preemptable `__stack_chk_guard` with static relocation model, we could emit
```
        ldr     r0, .LCPI0_0   // from LDRLIT_ga_abs
        ldr     r0, [r0]
        ldr     r0, [r0]
...
.LCPI0_0:
        .long   __stack_chk_guard@GOT
```

Unfortunately, such a GOT-generating relocation with computation `GOT(S) + A` does not exist for AArch32.
(The integrated assembler and GNU assembler have an issue that `@GOT` is not rejected)

Therefore, we have to use LDRLIT_ga_pcrel and expand to the PIC+GOT code sequence:
```
.LCPI0_0:
.Ltmp0:
        .long   __stack_chk_guard(GOT_PREL)-((.LPC0_0+8)-.Ltmp0)
```

Comment (efriedma-quic) [reply llvm/lib/Target/ARM/ARMInstrInfo.cpp:107]: Oh, it's also a relocation thing... that makes sense.  Could you add a comment in the code explaining this so it's clear why we're using this specific path on ELF?

Maybe for clarity, also compute "do we need to force a got-pic sequence" as a boolean, then make the inner check something like `if (TM.isPositionIndependent() || ForceELFGOTPIC`)`.

Comment (MaskRay) [reply llvm/lib/Target/ARM/ARMInstrInfo.cpp:107]: Added a comment.

Edited my previous comment: `R_ARM_GOT_ABS` is present, but it does not have assembler support.

**Comment by MaskRay** [comment]
I have tested that this patch does not change code generation for the following configurations:
https://gist.github.com/MaskRay/927d715e190617f8d8a2a066893705e3
```
#!/bin/zsh
CC=/tmp/Rel/bin/clang
if [[ -n $1 ]]; then
  update=1
  echo updated
else
  update=
fi

[[ ! -e a.cc ]] && echo 'void bar(char *); void foo() { char a[200]; bar(a); }' > a.cc

run() {
  if [[ -n $update ]]; then
    $CC "${@:2}" -S -fstack-protector a.cc -o $1.s
  else
    $CC "${@:2}" -S -fstack-protector a.cc -o - | diff -u $1.s -
  fi
}

run arm.linux.nopic --target=arm-linux-gnueabi -fno-pic
run arm.linux.pie --target=arm-linux-gnueabi -fpie
run arm.linux.pic --target=arm-linux-gnueabi -fpic
run armv6.darwin.nopic --target=armv6-apple-darwin -fno-pic
run armv6.darwin.dynamicnopic --target=armv6-apple-darwin -mdynamic-no-pic
run armv6.darwin.pic --target=armv6-apple-darwin -fpic
run armv7.darwin.nopic --target=armv7-apple-darwin -mcpu=cortex-a8 -fno-pic
run armv7.darwin.dynamicnopic --target=armv7-apple-darwin -mcpu=cortex-a8 -mdynamic-no-pic
run armv7.darwin.pic --target=armv7-apple-darwin -mcpu=cortex-a8 -fpic
run arm64.darwin.pic --target=arm64-apple-darwin
```

File: llvm/lib/CodeGen/TargetLoweringBase.cpp [hunk 2]
```diff
@@ -2019,7 +2019,9 @@ void TargetLoweringBase::insertSSPDeclarations(Module &M) const {
     if (M.getDirectAccessExternalData() &&
         !TM.getTargetTriple().isWindowsGNUEnvironment() &&
         !TM.getTargetTriple().isOSFreeBSD() &&
-        !TM.getTargetTriple().isOSDarwin())
+        (!TM.getTargetTriple().isOSDarwin() ||
+         (TM.getTargetTriple().isARM() &&
```
Comment (MaskRay) [inline llvm/lib/CodeGen/TargetLoweringBase.cpp:2023]: This ensures that code generation for arm*-apple-darwin -fno-pic does not change.

If `arm*-apple-darwin -fno-pic` is ok to use indirection, we can clean up this afterwards.

**Comment by nickdesaulniers** [comment]
> This patch fixes https://github.com/llvm/llvm-project/issues/6499 while not changing behavior for the following configurations:

Is that the right link?

**Comment by MaskRay** [comment]
> > This patch fixes #6499 while not changing behavior for the following configurations:
> 
> Is that the right link?

Thanks for catching my typo. Fixed to #64999

File: llvm/lib/CodeGen/TargetLoweringBase.cpp (hunk 2 above)
Comment (efriedma-quic) [reply llvm/lib/CodeGen/TargetLoweringBase.cpp:2023]: Does the current version of the patch actually change the logic for MachO targets outside of this change?  I can't find any logical differences (the ARMInstrInfo.cpp changes are guarded by isTargetELF(), and the ARMBaseInstrInfo.cpp change is unreachable due to an earlier isTargetMachO()).

Comment (MaskRay) [reply llvm/lib/CodeGen/TargetLoweringBase.cpp]: I believe this patch does not change Mach-O behavior.

This `llvm/lib/CodeGen/TargetLoweringBase.cpp` change can also be removed and the behavior does not change. Sorry for including it here. However, the file is somewhat related.

I've added a commit to this PR and this now becomes
```
    if (M.getDirectAccessExternalData() &&
        !TM.getTargetTriple().isWindowsGNUEnvironment() &&
        !TM.getTargetTriple().isOSFreeBSD() &&
        (!TM.getTargetTriple().isOSDarwin() ||
         TM.getRelocationModel() == Reloc::Static))
      GV->setDSOLocal(true);
  }
```

This `TM.getRelocationModel() == Reloc::Static` condition refines 1ec30106a557b2c63d6d6c6b93634674aa65cdb9 .

I think Mach-O's static relocation model largely ignores `dso_local` IR markers and relies on `TM.shouldAssumeDSOLocal` to return true for static relocation code model. However, it seems more appropriate to me to encode this information correctly, which can probably enable code share between ELF and Mach-O.

**Review by efriedma-quic — APPROVED** [review]
LGTM

**Comment by MaskRay** [comment]
Rebase after test adjustment at 6ae7b735dbd50eb7ade1573a86d037a2943e679c
```
    [ARM][test] Improve stack-protector tests

    llvm/test/LTO/ARM/ssp-static-reloc.ll is more about using the static
    relocation model with "PIC Level" and unrelated to the LTO infrastructure.
    Move the test. Update stack_guard_remat.ll to clearly test "PIC Level"
    with the relevant relocation models.

 llvm/test/CodeGen/ARM/stack-guard-elf.ll   | 41 +++++++++++++++++++++++++++++++++++++++++
 llvm/test/CodeGen/ARM/stack_guard_remat.ll | 16 ++++++++++------
 llvm/test/LTO/ARM/ssp-static-reloc.ll      | 40 ----------------------------------------
 3 files changed, 51 insertions(+), 46 deletions(-)
```

**Pull request merged**